File: LeetCode/0726-number-of-atoms/0726-number-of-atoms.cpp
class Solution {
public:
    string countOfAtoms(string formula) {
        stack<unordered_map<string, int>> st;
        int n = formula.size();
        int i = 0;
        st.push(unordered_map<string, int>());
        
        while (i < n) {
            if (formula[i] == '(') {
                st.push(unordered_map<string, int>());
                i++;
            } else if (formula[i] == ')') {
                unordered_map<string, int> curr = st.top();
                st.pop();
                i++;
                int mul = 0;
                while (i < n && isdigit(formula[i])) {
                    mul = mul * 10 + (formula[i] - '0');
                    i++;
                }
                if (mul == 0) mul = 1;
                
                for (auto &x : curr) {
                    curr[x.first] *= mul;
                }
                
                for (auto &x : curr) {
                    st.top()[x.first] += x.second;
                }
            } else {
                string currElement;
                currElement += formula[i++];
                while (i < n && islower(formula[i])) {
                    currElement += formula[i++];
                }
                int count = 0;
                while (i < n && isdigit(formula[i])) {
                    count = count * 10 + (formula[i] - '0');
                    i++;
                }
                if (count == 0) count = 1;
                st.top()[currElement] += count;
            }
        }
        
        string ans = "";
        map<string, int> mp(st.top().begin(), st.top().end());
        for (auto &x : mp) {
            ans += x.first;
            if (x.second > 1) {
                ans += to_string(x.second);
            }
        }
        return ans;
    }
};

File: LeetCode/0726-number-of-atoms/0726-number-of-atoms.py
class Solution:
    def countOfAtoms(self, formula: str) -> str:
        stack = []
        n = len(formula)
        i = 0
        stack.append(collections.defaultdict(int))
        
        while i < n:
            if formula[i] == '(':
                stack.append(collections.defaultdict(int))
                i += 1
            elif formula[i] == ')':
                curr = stack.pop()
                i += 1
                mul = 0
                while i < n and formula[i].isdigit():
                    mul = mul * 10 + int(formula[i])
                    i += 1
                if mul == 0:
                    mul = 1
                
                for elem, cnt in curr.items():
                    curr[elem] *= mul
                
                for elem, cnt in curr.items():
                    stack[-1][elem] += cnt
            else:
                curr_element = formula[i]
                i += 1
                while i < n and formula[i].islower():
                    curr_element += formula[i]
                    i += 1
                count = 0
                while i < n and formula[i].isdigit():
                    count = count * 10 + int(formula[i])
                    i += 1
                if count == 0:
                    count = 1
                stack[-1][curr_element] += count
        
        ans = ""
        sorted_elements = sorted(stack[-1].items())
        for elem, cnt in sorted_elements:
            ans += elem
            if cnt > 1:
                ans += str(cnt)
        
        return ans